src/passes/OptimizeInstructions.cpp

```diff
@@ -594,25 +594,77 @@ struct OptimizeInstructions
-        if (auto* left = binary->left->dynCast<Unary>()) {
-          if (left->op == EqZInt32) {
-            if (auto* right = binary->right->dynCast<Unary>()) {
-              if (right->op == EqZInt32) {
-                // reuse one unary, drop the other
-                auto* leftValue = left->value;
-                left->value = binary;
-                binary->left = leftValue;
-                binary->right = right->value;
-                binary->op = OrInt32;
-                return left;
+        if (binary->op == AndInt32) {
+          if (auto* left = binary->left->dynCast<Unary>()) {
+            if (left->op == EqZInt32) {
+              if (auto* right = binary->right->dynCast<Unary>()) {
+                if (right->op == EqZInt32) {
+                  // reuse one unary, drop the other
+                  auto* leftValue = left->value;
+                  left->value = binary;
+                  binary->left = leftValue;
+                  binary->right = right->value;
+                  binary->op = OrInt32;
+                  return left;
+                }
+              }
+            }
+          }
+        }
+        // (x ^ -1) & x   ==>    0
+        // (x ^ -1) & (y ^ -1)   ==>    (x | y) ^ -1
+        if (auto* left = binary->left->dynCast<Binary>()) {
+          if (left->op == Abstract::getBinary(binary->type, Abstract::Xor)) {
+            if (auto* c = left->right->dynCast<Const>()) {
+              if (c->value.getInteger() == -1LL) {
+                // (x ^ -1) & x   ==>    0
+                if (ExpressionAnalyzer::equal(left->left, binary->right) &&
+                    !EffectAnalyzer(getPassOptions(), features, binary->right)
+                       .hasSideEffects()) {
+                  return LiteralUtils::makeZero(binary->type, *getModule());
+                }
+                // (x ^ -1) & (y ^ -1)   ==>    (x | y) ^ -1
+                if (auto* right = binary->right->dynCast<Binary>()) {
+                  if (left->type == right->type && left->type == binary->type) {
+                    if (right->op ==
+                        Abstract::getBinary(binary->type, Abstract::Xor)) {
+                      if (auto* c = right->right->dynCast<Const>()) {
+                        if (c->value.getInteger() == -1LL) {
+                          // reuse one binary, drop the other
+                          auto* leftValue = left->left;
+                          left->left = binary;
+                          binary->left = leftValue;
+                          binary->right = right->left;
+                          binary->op =
+                            Abstract::getBinary(binary->type, Abstract::Or);
+                          return left;
+                        }
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+        // x & (x ^ -1)   ==>    0
```

Comment:
this should not be needed at least for `x` being a `local.get` due to the current canonicalization. instead of duplicating code with flipped things like this, I think we should try to make canonicalization handle it. one possibility is to canonicalize to put larger things on the left.

alternatively, we could make a helper function that gets the left and the right, and has all the logic, and we call it twice (once flipped).

both might be separated out from this PR perhaps.

Reply:
In discord I proposed hoisting (swapping) subexpressions which contain constants to the left. It will always collect all consts in rhs even for sub-expressions. I could create new PR for this, or did you want to do it yourself?

Reply:
Sorry, I missed that on discord.

When you say "which contain constants" did you mean to look in the entire IR tree there? I hope there's a way we can look into a fixed number of levels deep, to avoid polynomial times in bad cases.

Reply:
Yeah, just look in one level deep like:
`((X op1 C) op2 Y)` -> `(Y op2 (X op1 C))`, is `op2` is commutative and `C` is a constant

Reply:
Made PR for this: https://github.com/WebAssembly/binaryen/pull/2894

Reply:
I still think a helper function would be good here. It would remove some code duplication.